Kt 56387 sealed classes improvement (#3989)
JetBrains/kotlin-web-site: docs/topics/sealed-classes.md

[//]: # (title: Sealed classes and interfaces)

_Sealed_ classes and interfaces represent restricted class hierarchies that provide more control over inheritance.

> Sealed classes and interfaces, in contrast to `internal` classes, ensure inherent exhaustiveness in `when` expressions within their declared module or package.

Reviewer: I don't think this is in contrast to `internal` classes, but to `open` classes.

```suggestion
Sealed classes and interfaces, in contrast to `open` ones (which is the default), ensure inherent exhaustiveness in `when` expressions.
```

All direct subclasses of a sealed class are known at compile time. No other subclasses may appear outside

> the module and package within which the sealed class is defined. 

Reviewer: Actually, I'm not sure about having to be in the same package (but definitely they are all in the same module).

>
{type="tip"}

## Declare a sealed class or interface

To declare a sealed class or interface, use the `sealed` modifier:

```kotlin
// Create a sealed interface
sealed interface Error

// Create a sealed class
sealed class IOError(): Error

// Define subclasses that extend sealed class 'IOError'
class FileReadError(val file: File): IOError()
class DatabaseError(val source: DataSource): IOError()

// Create a singleton object implementing the 'Error' sealed interface 
object RuntimeError : Error
```

This example could represent a library's API that contains error classes to let library users handle errors that it can throw. 
If the hierarchy of such error classes includes interfaces or abstract classes visible in the public API, then nothing 
prevents other developers from implementing or extending them in the client code. 
Since the library doesn't know about errors declared outside of it, it can’t treat them consistently with its own classes. 
However, with a **sealed** hierarchy of error classes, library authors can be sure that they know all the possible error 

> types and that other ones can’t appear later.

Reviewer 2: ```suggestion
types and that other error types can't appear later.
```

The hierarchy of the example looks like this:

> ![Hierarchy Illustration of Sealed Classes and Interfaces](sealedClasses_interfaces.svg){width=700}

Author: as discussed - I'm changing the name to lowercase+dashes and creating a new folder for concepts

sealed class IOError(): Error

// Open class 'CustomError' extends 'Error' and can be extended anywhere it's visible
open class CustomError(): Error
```

### Inheritance in multiplatform projects

There is one more inheritance restriction in [multiplatform projects](multiplatform-get-started.md): direct subclasses of sealed classes must
reside in the same [source set](multiplatform-discover-project.md#source-sets). It applies to sealed classes without the [expected and actual modifiers](multiplatform-expect-actual.md).

If a sealed class is declared as `expect` in a common source set and have `actual` implementations in platform source sets,
both `expect` and `actual` versions can have subclasses in their source sets. Moreover, if you use a hierarchical structure,
you can create subclasses in any source set between the `expect` and `actual` declarations. 

[Learn more about the hierarchical structure of multiplatform projects](multiplatform-hierarchy.md). 

## Use sealed classes with when expression

The key benefit of using sealed classes comes into play when you use them in a [`when`](control-flow.md#when-expression)
expression.

> The `when` expression, used with a sealed class, allows the Kotlin compiler to check exhaustively that all possible cases are covered. 

Author: 🤔 - That's a tough one. I like all 3 (the original included).
In my opinion:
1. The original is a bit wordy but conveys all the information. 
2. "The when expression with a sealed class..." It's more concise, but I feel it's also a bit more passive - it might imply that the cases are covered by default, rather than being actively checked by the compiler.
3. "If you use.." I think this one is a bit less direct in its structure, but a bit more active than the original.  

I don't think it's an issue being a bit more wordy in this case to belabor a point, because it's one of the most important functionalities of sealed classes.

> The `when` expression, used with a sealed class, allows the Kotlin compiler to check exhaustively that all possible cases are covered. 

Reviewer 3: Fair enough. Up to you.

In such cases, you don't need to add an `else` clause. Here's an example demonstrating this:

```kotlin
// Sealed class and its subclasses
sealed class Error {
    class FileReadError(val file: String): Error()
    class DatabaseError(val source: String): Error()
    object RuntimeError : Error()
}

//sampleStart
// Function to log errors
fun log(e: Error) = when(e) {
    is Error.FileReadError -> println("Error while reading file ${e.file}")
    is Error.DatabaseError -> println("Error while reading from database ${e.source}")
    Error.RuntimeError -> println("Runtime error")
    // No `else` clause is required because all the cases are covered
}
//sampleEnd

// List all errors
fun main() {
    val errors = listOf(
        Error.FileReadError("example.txt"),
        Error.DatabaseError("usersDatabase"),
        Error.RuntimeError
    )

    errors.forEach { log(it) }
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.5"}

> In multiplatform projects, if you have a sealed class with a `when` expression as an 

> > [expected declarations](multiplatform-expect-actual.md) in your common code, you still need an `else` branch. 

Reviewer 2: ```suggestion
> [expected declaration](multiplatform-expect-actual.md) in your common code, you still need an `else` branch. 
```

> This is because subclasses of `actual` platform implementations may extend sealed classes that 
> aren't known in the common code.
>
{type="note"}

## Use case scenarios

> Let's explore some practical scenarios where the application of sealed classes or interfaces can be particularly helpful.

Reviewer 3: Let's simplify the second part of the sentence a bit:

"... where sealed classes and interfaces can be particularty helpful/useful"

Does it work?

Author: Yes, I think it sounds much better! - I'll update it

### State management in UI applications

> You can use sealed classes to represent different states of a UI in an application. 

Reviewer 3: different states of UI -> different/vaious UI states

WDYT?

This approach allows for structured and safe handling of UI changes. 

> This example demonstrates how to manage various UI states:

Reviewer 3: Do we need this sentence or it could be omited?

Author: I don't think it's necessary, somewhat eases the transition from explanation to example, but I don't think it's necessary.

Reviewer 3: Up to you.

```kotlin
sealed class UIState {
	data object Loading : UIState()
	data class Success(val data: String) : UIState()
	data class Error(val exception: Exception) : UIState()
}

fun updateUI(state: UiState) {
	when (state) {
    	is UIState.Loading -> showLoadingIndicator()
    	is UIState.Success -> showData(state.data)
    	is UIState.Error -> showError(state.exception)
	}
}
```

### Payment method handling

In practical business applications, handling various payment methods efficiently is a common requirement. 

> This example illustrates how to implement such business logic using sealed classes. 

Reviewer 3: This example illustrates how to implement such business logic using sealed classes. ->

You can use sealed classes with `when` expressions to implement such business logic.

Author: I really like this suggestion - more active and puts more emphasis on the when expression. - I'll update it

By representing different payment methods as subclasses of a sealed class, it establishes a clear and manageable 
structure for processing transactions:

```kotlin
sealed class Payment {
	data class CreditCard(val number: String, val expiryDate: String) : Payment()
	data class PayPal(val email: String) : Payment()
	data object Cash : Payment()
}

fun processPayment(payment: Payment) {
	when (payment) {
    	is Payment.CreditCard -> processCreditCardPayment(payment.number, payment.expiryDate)
    	is Payment.PayPal -> processPayPalPayment(payment.email)
    	is Payment.Cash -> processCashPayment()
	}
}
```

> `Payment` is a sealed class that represents different payment methods in an e-commerce system: 

Reviewer 3: Maybe organize these senteces into a list?

* `Payment` ...
* `processPayment()` ...

WDYT?

Author: hmm 🤔 I think adding a line break between the processPayment() sentence and the previous ones would aid with readability a bit more. 

(Converting this to a list would look something like this (I think it looks a bit awkward):

* `Payment`: a sealed class that represents different payment methods in an e-commerce system:
  * ˙CreditCard˙
  * ˙Paypal˙
  * ˙Cash˙
* ˙processPayment()˙: a function that demonstrates how to handle different payment methods.

Reviewer 3: WFM!

`CreditCard`, `PayPal`, and `Cash`. Each subclass can have its specific properties, like `number` and `expiryDate` for 
`CreditCard`, and `email` for `PayPal`. The `processPayment()` function demonstrates how to handle different payment methods.
This approach ensures that all possible payment types are considered, and the system remains flexible for new payment 
methods to be added in the future.

### API request-response handling

> This example explores a practical implementation of a user authentication system that handles API requests and responses 

Author: I like it it's a bit more active - I'll update it (will also add info about sealed interfaces)

To declare a sealed class or interface, use the `sealed` modifier:

```kotlin
// Create a sealed interface
sealed interface Error

// Create a sealed class that implements sealed interface Error
sealed class IOError(): Error

// Define subclasses that extend sealed class 'IOError'
class FileReadError(val file: File): IOError()
class DatabaseError(val source: DataSource): IOError()

// Create a singleton object implementing the 'Error' sealed interface 
object RuntimeError : Error
```

This example could represent a library's API that contains error classes to let library users handle errors that it can throw. 
If the hierarchy of such error classes includes interfaces or abstract classes visible in the public API, then nothing 
prevents other developers from implementing or extending them in the client code. 
Since the library doesn't know about errors declared outside of it, it can’t treat them consistently with its own classes. 
However, with a **sealed** hierarchy of error classes, library authors can be sure that they know all the possible error 
types and that other error types can't appear later.

The hierarchy of the example looks like this:

> ![Hierarchy Illustration of Sealed Classes and Interfaces](sealed-classes-interfaces.svg){width=700}

Reviewer 3: ```suggestion
![Hierarchy illustration of sealed classes and interfaces](sealed-classes-interfaces.svg){width=700}
```

    // A sealed class constructor has protected visibility by default. It's visible inside this class and its subclasses 
    constructor() { /*...*/ }

    // Private constructor, visible inside this class only. 
    // Using a private constructor in a sealed class allows for even stricter control over instantiation, enabling specific initialization procedures within the class.
    private constructor(description: String): this() { /*...*/ }

    // This will raise an error because public and internal constructors are not allowed in sealed classes
    // public constructor(code: Int): this() {} 
}
```

## Inheritance

Direct subclasses of sealed classes and interfaces must be declared in the same package. They may be top-level or nested
inside any number of other named classes, named interfaces, or named objects. Subclasses can have any [visibility](visibility-modifiers.md)
as long as they are compatible with normal inheritance rules in Kotlin.

Subclasses of sealed classes must have a properly qualified name. They can't be local or anonymous objects.

> `enum` classes can't extend a sealed class, or any other class. However, they can implement sealed interfaces:
>

> > ```kotlin
> > sealed interface Error
> > 
> > // enum class extending the sealed interface Error
> > enum class ErrorType : Error {
> >    FILE_ERROR, DATABASE_ERROR
> > }

Reviewer 3: ```suggestion
> ```kotlin
> sealed interface Error
> 
> // enum class extending the sealed interface Error
> enum class ErrorType : Error {
>     FILE_ERROR, DATABASE_ERROR
> }
```

>
> ```
> 
{type="note"}

These restrictions don't apply to indirect subclasses. If a direct subclass of a sealed class is not marked as sealed,
it can be extended in any way that its modifiers allow:

```kotlin

> // Sealed interface 'Error'has implementations only in the same package and module

Reviewer 3: ```suggestion
// Sealed interface 'Error' has implementations only in the same package and module
```

object RuntimeError : Error
```

The hierarchy of our sealed interface and sealed class example looks like this:
![Hierarchy Illustration of Sealed Classes and Interfaces](sealedClasses_interfaces.png){width=600}

> A sealed class itself is an [abstract](classes.md#abstract-classes) and cannot be instantiated directly. 

Reviewer: ```suggestion
A sealed class itself is always [abstract](classes.md#abstract-classes) and, as a result, cannot be instantiated directly. 
```

However, it may contain or inherit constructors. These constructors are not for creating instances of the sealed class itself but for its subclasses.
Let's look at an example where we have a sealed class called `Error` with several subclasses, which we instantiate:

```kotlin
sealed class Error(val message: String) {
    class NetworkError : Error("Network failure")
    class DatabaseError : Error("Database cannot be reached")
    class UnknownError : Error("An unknown error has occurred")
}

fun main() {
    val errors = listOf(Error.NetworkError(), Error.DatabaseError(), Error.UnknownError())
    errors.forEach { println(it.message) }
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.5"}

> In some sense, sealed classes are similar to [`enum`](enum-classes.md) classes: the set of values
> for an enum type is also restricted, but each enum constant exists only as a _single instance_, whereas a subclass
> of a sealed class can have _multiple_ instances, each with its own state. 

Reviewer: ```suggestion
Sealed classes share with [`enum`](enum-classes.md) classes a restriction over the set of values.
Note that each enum constant exist only as a _single instance_,
whereas subclasses of a sealed class may have _multiple_ instance, each with its own state.
```

## Sealed classes and when expression

The key benefit of using sealed classes comes into play when you use them in a [`when`](control-flow.md#when-expression)
expression.
The `when` expression, used with a sealed class, allows the Kotlin compiler to check exhaustively that all possible cases are covered. In such cases, you don't need to add an `else` clause. 
Here's an example demonstrating this with a `when` expression:

```kotlin
// Sealed class and its subclasses
sealed class Error
class FileReadError(val file: String): Error()
class DatabaseError(val source: String): Error()

> object RuntimeError : Error()

Reviewer: You are switching styles here: before you defined things as nested classes. Not super-important, but may puzzle people who are finding this notion for the first time.

Author: True! - I agree - I'll make this one nested as well to stay consistent

//sampleStart
// Function to log errors
fun log(e: Error) = when(e) {
    is FileReadError -> println("Error while reading file ${e.file}")
    is DatabaseError -> println("Error while reading from database ${e.source}")

>     is RuntimeError -> println("Runtime error")

Reviewer: By convention you use only the name of the object

```suggestion
    RuntimeError -> println("Runtime error")
```

Author: thank you! - very nice feedback - I'll implement it right away
